Lib/xml/etree/ElementTree.py

```diff
@@ -836,63 +836,81 @@ def _get_writer(file_or_filename, encoding):
+    # maps qnames to *encoded* prefix:local names,
+    # value is a pair of prefix:local strings,
+    # second value is for attribute names
+    qnames = {None: (None, None)}
 
     # maps uri:s to prefixes
     namespaces = {}
-    if default_namespace:
-        namespaces[default_namespace] = ""
 
-    def add_qname(qname):
+    def serialize_qname(qname, is_attr):
         # calculate serialized qname representation
-        try:
-            if qname[:1] == "{":
-                uri, tag = qname[1:].rsplit("}", 1)
+        if qname[:1] == "{":
+            uri, local = qname[1:].rsplit("}", 1)
+            if not is_attr and uri == default_namespace:
+                prefix = ""
+            else:
                 prefix = namespaces.get(uri)
                 if prefix is None:
                     prefix = _namespace_map.get(uri)
                     if prefix is None:
                         prefix = "ns%d" % len(namespaces)
                     if prefix != "xml":
                         namespaces[uri] = prefix
-                if prefix:
-                    qnames[qname] = "%s:%s" % (prefix, tag)
-                else:
-                    qnames[qname] = tag # default element
+            if prefix:
+                return "%s:%s" % (prefix, local)
             else:
-                if default_namespace:
-                    # FIXME: can this be handled in XML 1.0?
-                    raise ValueError(
-                        "cannot use non-qualified names with "
-                        "default_namespace option"
-                        )
-                qnames[qname] = qname
-        except TypeError:
-            _raise_serialization_error(qname)
+                return local # default element
+        else:
+            if not is_attr and default_namespace:
+                # FIXME: can this be handled in XML 1.0?
+                raise ValueError(
+                    "cannot use non-qualified name (<%s>) with "
+                    "default_namespace option" % qname
+                    )
+            return qname
+
+    def add_qname(qname, is_attr=False):
+        ser_tag, ser_attr = qnames.get(qname, (None, None))
+        if is_attr:
+            if ser_attr is None:
+                ser_attr = serialize_qname(qname, True)
+                if not default_namespace:
+                    ser_tag = ser_attr
+                qnames[qname] = (ser_tag, ser_attr)
+        else:
+            if ser_tag is None:
+                ser_tag = serialize_qname(qname, False)
+                if not default_namespace:
+                    ser_attr = ser_tag
+                qnames[qname] = (ser_tag, ser_attr)
 
     # populate qname and namespaces table
     for elem in elem.iter():
         tag = elem.tag
         if isinstance(tag, QName):
-            if tag.text not in qnames:
-                add_qname(tag.text)
+            add_qname(tag.text)
         elif isinstance(tag, str):
-            if tag not in qnames:
```

Comment:
These conditions might have been there for performance reasons. Not sure if it matters, but that's up to some benchmarking I think.